src/main/java/org/opensearch/knn/index/query/KNNWeight.java

```diff
@@ -441,14 +445,72 @@ private boolean isExactSearchThresholdSettingSet(int filterThresholdValue) {
+     * @param context LeafReaderContext
+     * @param filterIdsCount count of filtered Doc ids
+     * @param annResultCount Count of Nearest Neighbours we got after doing filtered ANN Search.
+     * @return boolean - true if exactSearch needs to be done after ANNSearch.
+     */
+    private boolean canDoExactSearch(final LeafReaderContext context, final int filterIdsCount, final int annResultCount) {
+        if (annResultCount == 0 && isMissingNativeEngineFiles(context)) {
+            log.info("Perform exact search after approximate search since no native engine files are available");
+            return true;
+        }
+        if (canDoFilteredExactSearchAfterANNSearch(filterIdsCount, annResultCount)) {
+            log.debug(
+                "Doing ExactSearch after doing ANNSearch as the number of documents returned are less than "
+                    + "K, even when we have more than K filtered Ids. K: {}, ANNResults: {}, filteredIdCount: {}",
+                this.knnQuery.getK(),
+                annResultCount,
+                filterIdsCount
+            );
+            return true;
+        }
+        return false;
+    }
+
     /**
      * This condition mainly checks during filtered search we have more than K elements in filterIds but the ANN
-     * doesn't yeild K nearest neighbors.
+     * doesn't yield K nearest neighbors.
      * @param filterIdsCount count of filtered Doc ids
      * @param annResultCount Count of Nearest Neighbours we got after doing filtered ANN Search.
      * @return boolean - true if exactSearch needs to be done after ANNSearch.
      */
-    private boolean canDoExactSearchAfterANNSearch(final int filterIdsCount, final int annResultCount) {
+    private boolean canDoFilteredExactSearchAfterANNSearch(final int filterIdsCount, final int annResultCount) {
         return filterWeight != null && filterIdsCount >= knnQuery.getK() && knnQuery.getK() > annResultCount;
     }
+
+    /**
+     * This condition mainly checks whether segments has native engine files or not
+     * @return boolean - false if exactSearch needs to be done since no native engine files are in segments.
+     */
+    private boolean isMissingNativeEngineFiles(LeafReaderContext context) {
+        final SegmentReader reader = Lucene.segmentReader(context.reader());
+        final FieldInfo fieldInfo = reader.getFieldInfos().fieldInfo(knnQuery.getField());
+        // if segment has no documents with at least 1 vector field, field info will be null
+        if (fieldInfo == null) {
+            return false;
+        }
+        final KNNEngine knnEngine = extractKNNEngine(fieldInfo);
+        final List<String> engineFiles = KNNCodecUtil.getEngineFiles(
+            knnEngine.getExtension(),
+            knnQuery.getField(),
+            reader.getSegmentInfo().info
+        );
+        return engineFiles.isEmpty();
+    }
+
+    private KNNEngine extractKNNEngine(FieldInfo fieldInfo) {
```

Comment:
there is already a class called as FieldInfoExtracter and this extractKNNEngine function is present there, use that funciton.

Reply:
Ack